Foundation/NSURL.swift

```diff
@@ -638,13 +665,464 @@ open class NSURL : NSObject, NSSecureCoding, NSCopying {
     override open var _cfTypeID: CFTypeID {
         return CFURLGetTypeID()
     }
+    
+    private func _initResourceInfo() {
+        let pInfo = UnsafeMutablePointer<NSMutableArray>.allocate(capacity: 1)
+        pInfo.initialize(to: NSMutableArray())
+        _CFSwiftRetain(UnsafeMutableRawPointer(pInfo)) // Balanced with resourceInfo release in _CFURLDeallocate()
+        pInfo.pointee.add(NSLock())
+        pInfo.pointee.add(NSMutableDictionary())
+        __CFURLSetResourceInfoPtr(_cfObject, UnsafeMutableRawPointer(pInfo))
+    }
+    
+    private func _accessResourceDictionary<T>(_ handler: (_ resourcesCache: NSMutableDictionary) -> T) -> T? {
+        // Resource values are only accessible for file urls.
+        guard _CFURLIsFileURL(_cfObject) else {
+            return nil
+        }
+        if __CFURLResourceInfoPtr(_cfObject) == nil {
+            _initResourceInfo()
+        }
+        guard let resourceInfoPtr = __CFURLResourceInfoPtr(_cfObject)?.assumingMemoryBound(to: NSMutableArray.self),
+            let lock = resourceInfoPtr.pointee.object(at: 0) as? NSLock,
+            let resourcesCache = resourceInfoPtr.pointee.object(at: 1) as? NSMutableDictionary else {
+            return nil
+        }
+        lock.lock()
+        defer {
+            lock.unlock()
+        }
+        return handler(resourcesCache)
+        
+    }
+    
+    open func removeAllCachedResourceValues() {
+        _accessResourceDictionary { (resourcesCache) in
+            resourcesCache.removeAllObjects()
+        }
+    }
+    
+    open func removeCachedResourceValue(forKey key: URLResourceKey) {
+        _accessResourceDictionary { (resourcesCache) in
+            resourcesCache.removeObject(forKey: key.rawValue)
+        }
+    }
+    
+    open func resourceValues(forKeys keys: [URLResourceKey]) throws -> [URLResourceKey : Any] {
+        
+        func hasFlag(path: String, flag: Int32) throws -> Bool {
+            #if canImport(Darwin) || os(FreeBSD)
+            let stat = try FileManager.default._lstatFile(atPath: path)
+            return stat.st_flags & UInt32(flag) != 0
+            #elseif canImport(Glibc)
+            let fd = open(path, O_RDWR)
```

Comment:
`path` needs to be converted using `_fileSystemRepresentation` but it might actually be better to replace the `open()` with something like:
```
let fh = try FileHandle(..)
var attr: Int32 = 0
ioctl(fh.fileDescriptor, FS_IOC_GETFLAGS, &attr)
return attrs & flag != 0
```